# Homework 4: Data Transformation with dplyr - Part 2 (Mutate, Summarize, Group_by, Count)

**Student Name:** Michael Alexander

**Due Date:** 09/21/2025

Welcome to Homework 4! This assignment builds on your foundational dplyr skills by introducing advanced data transformation techniques essential for business analytics.

## Learning Objectives

By completing this homework, you will:
- **Master mutate()**: Create new variables and calculated fields for business insights
- **Apply summarize()**: Generate aggregate statistics and key performance indicators (KPIs)
- **Utilize group_by()**: Perform grouped analysis across business dimensions
- **Implement count()**: Conduct frequency analysis and cross-tabulations
- **Develop business intelligence**: Combine functions to create comprehensive analytics

## Business Context

You'll be working with real company sales data to perform the type of analysis that drives business decisions. This includes:
- **Financial Analysis**: Calculate profit margins, ROI, and efficiency metrics
- **Performance Segmentation**: Categorize transactions and customers by performance
- **Regional Analysis**: Compare performance across geographic regions
- **Product Analysis**: Evaluate product category performance
- **Time-based Analysis**: Identify trends and seasonal patterns
- **Sales Rep Performance**: Assess individual representative effectiveness

## Dataset Information

You'll be working with `company_sales_data.csv` which contains:
- **Sales transactions** with revenue, cost, and unit data
- **Geographic information** with regional breakdowns
- **Product categories** for portfolio analysis
- **Sales representative** performance data
- **Date information** for trend analysis
- **Customer metrics** for segmentation analysis

## Instructions

Complete each section by writing R code in the designated areas. Focus on creating clean, well-commented code that demonstrates your understanding of both the technical concepts and business applications.

**Remember**: In real business analytics, you're not just manipulating data - you're uncovering insights that drive strategic decisions!

## Part 1: Setup and Data Import

**Task**: Load the necessary packages and import your dataset.

**Business Context**: Every analysis begins with proper data setup. In professional environments, data scientists spend significant time ensuring data is properly loaded and validated before analysis begins.

**What you need to do**:
1. Load the `tidyverse` package (which includes dplyr)
2. Import the `company_sales_data.csv` file 
3. Examine the structure and basic properties of your data
4. Perform initial data validation

**Hint**: Use `read_csv()` for better data type detection, and always examine your data structure with `str()`, `summary()`, and `head()` functions.

In [19]:
# Load required packages
# TODO: Load tidyverse
library(tidyverse)

# Import the company sales data
# TODO: Read the company_sales_data.csv file into a variable called company_data
company_data <- read_csv("company_sales_data.csv")

# Examine the dataset structure
# TODO: Use str() to examine the data structure
str(company_data)

# Display the first few rows
# TODO: Use head() to show the first 6 rows
head(company_data, 6)

# Generate summary statistics
# TODO: Use summary() to get basic statistics for all columns
summary(company_data)

# Check dataset dimensions
# TODO: Display the number of rows and columns using nrow() and ncol()
cat("Dataset dimensions:", nrow(company_data), "rows x", ncol(company_data), "columns\n")

Rows: 300 Columns: 8
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr  (3): Sales_Rep_Name, Region, Product_Category
dbl  (4): TransactionID, Revenue, Cost, Units_Sold
date (1): Sale_Date
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr  (3): Sales_Rep_Name, Region, Product_Category
dbl  (4): TransactionID, Revenue, Cost, Units_Sold
date (1): Sale_Date

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


spc_tbl_ [300 × 8] (S3: spec_tbl_df/tbl_df/tbl/data.frame)
 $ TransactionID   : num [1:300] 1 2 3 4 5 6 7 8 9 10 ...
 $ Sales_Rep_Name  : chr [1:300] "Carol Davis" "Carol Davis" "Carol Davis" "Bob Smith" ...
 $ Region          : chr [1:300] "Latin America" "Europe" "Europe" "Europe" ...
 $ Product_Category: chr [1:300] "Services" "Hardware" "Services" "Hardware" ...
 $ Revenue         : num [1:300] 20751 32360 39268 28865 3932 ...
 $ Cost            : num [1:300] 12253 24595 23291 12429 1778 ...
 $ Units_Sold      : num [1:300] 78 13 34 90 63 26 25 1 20 15 ...
 $ Sale_Date       : Date[1:300], format: "2023-04-24" "2023-06-09" ...
 - attr(*, "spec")=
  .. cols(
  ..   TransactionID = col_double(),
  ..   Sales_Rep_Name = col_character(),
  ..   Region = col_character(),
  ..   Product_Category = col_character(),
  ..   Revenue = col_double(),
  ..   Cost = col_double(),
  ..   Units_Sold = col_double(),
  ..   Sale_Date = col_date(format = "")
  .. )
 - attr(*, "problems")=<externalptr

TransactionID Sales_Rep_Name Region        Product_Category Revenue  Cost    
1 1             Carol Davis    Latin America Services         20750.92 12253.36
2 2             Carol Davis    Europe        Hardware         32359.98 24595.20
3 3             Carol Davis    Europe        Services         39268.40 23291.09
4 4             Bob Smith      Europe        Hardware         28865.09 12428.74
5 5             Frank Miller   Latin America Software          3932.36  1778.18
6 6             Carol Davis    Latin America Services         48209.75 26052.04
  Units_Sold Sale_Date 
1 78         2023-04-24
2 13         2023-06-09
3 34         2023-03-25
4 90         2023-04-11
5 63         2023-08-26
6 26         2023-09-04

 TransactionID    Sales_Rep_Name        Region          Product_Category  
 Min.   :  1.00   Length:300         Length:300         Length:300        
 1st Qu.: 75.75   Class :character   Class :character   Class :character  
 Median :150.50   Mode  :character   Mode  :character   Mode  :character  
 Mean   :150.50                                                           
 3rd Qu.:225.25                                                           
 Max.   :300.00                                                           
    Revenue           Cost         Units_Sold       Sale_Date         
 Min.   : 1032   Min.   :  567   Min.   :  1.00   Min.   :2023-01-02  
 1st Qu.:15034   1st Qu.: 7216   1st Qu.: 28.75   1st Qu.:2023-04-07  
 Median :26062   Median :13956   Median : 56.00   Median :2023-07-05  
 Mean   :25906   Mean   :14547   Mean   : 53.90   Mean   :2023-07-07  
 3rd Qu.:37708   3rd Qu.:21714   3rd Qu.: 80.00   3rd Qu.:2023-10-08  
 Max.   :49956   Max.   :37811   Max.   :100.00  

Dataset dimensions: 300 rows x 8 columns


## Part 2: Creating New Variables with mutate()

**Task**: Use `mutate()` to create new calculated fields for business analysis.

**Business Context**: Raw data rarely tells the complete story. Business analysts must create derived metrics like profit margins, efficiency ratios, and performance categories to generate actionable insights.

**Variables to create**:
1. **Profit**: Revenue minus Cost
2. **Profit_Margin**: (Profit / Revenue) × 100
3. **Cost_Ratio**: (Cost / Revenue) × 100  
4. **Revenue_Per_Unit**: Revenue divided by Units_Sold
5. **Cost_Per_Unit**: Cost divided by Units_Sold
6. **ROI**: (Profit / Cost) × 100

**Hint**: Use the pipe operator (`%>%`) to chain multiple mutate operations together. Remember to create meaningful variable names that clearly indicate what each metric represents.

In [20]:
# Create basic financial metrics using mutate()
company_data <- company_data %>%
  mutate(
    # TODO: Create Profit (Revenue - Cost)
    Profit = Revenue - Cost,
    
    # TODO: Create Profit_Margin ((Profit / Revenue) * 100)
    Profit_Margin = (Profit / Revenue) * 100,
    
    # TODO: Create Cost_Ratio ((Cost / Revenue) * 100)
    Cost_Ratio = (Cost / Revenue) * 100,
    
    # TODO: Create Revenue_Per_Unit (Revenue / Units_Sold)
    Revenue_Per_Unit = Revenue / Units_Sold,
    
    # TODO: Create Cost_Per_Unit (Cost / Units_Sold)
    Cost_Per_Unit = Cost / Units_Sold,
    
    # TODO: Create ROI ((Profit / Cost) * 100)
    ROI = (Profit / Cost) * 100
  )

# Display the new financial metrics
# TODO: Select and display columns: Revenue, Cost, Profit, Profit_Margin, ROI
company_data %>%
  select(Revenue, Cost, Profit, Profit_Margin, ROI) %>%
  head(10)

Revenue  Cost     Profit   Profit_Margin ROI      
1  20750.92 12253.36  8497.56 40.95028       69.34882
2  32359.98 24595.20  7764.78 23.99501       31.57031
3  39268.40 23291.09 15977.31 40.68745       68.59838
4  28865.09 12428.74 16436.35 56.94197      132.24470
5   3932.36  1778.18  2154.18 54.78084      121.14522
6  48209.75 26052.04 22157.71 45.96106       85.05173
7  33055.35 20432.24 12623.11 38.18780       61.78035
8  49856.54 37118.63 12737.91 25.54913       34.31676
9  13477.13  8685.85  4791.28 35.55119       55.16190
10 21188.96 16757.06  4431.90 20.91608       26.44796

## Part 3: Creating Categorical Variables

**Task**: Create categorical variables for business segmentation using conditional logic.

**Business Context**: Segmentation is crucial for targeted business strategies. By categorizing transactions and customers, businesses can tailor their approaches to different performance levels and customer types.

**Categories to create**:
1. **Performance_Category**: Based on Profit_Margin (High: >50%, Medium: 30-50%, Low: <30%)
2. **Revenue_Size**: Based on Revenue (Large: >30000, Medium: 15000-30000, Small: <15000)
3. **Deal_Size**: Based on Units_Sold (Bulk: >50, Standard: 20-50, Small: <20)
4. **High_Value_Customer**: Flag for Revenue > 25000
5. **Profitable_Deal**: Flag for Profit_Margin > 40%

**Hint**: Use `case_when()` for multiple conditions or `ifelse()` for simple binary classifications.

In [21]:
# Create categorical variables for business segmentation
company_data <- company_data %>%
  mutate(
    # TODO: Create Performance_Category using case_when()
    # High: Profit_Margin > 50, Medium: Profit_Margin > 30, Low: everything else
    Performance_Category = case_when(
      Profit_Margin > 50 ~ "High",
      Profit_Margin > 30 ~ "Medium",
      TRUE ~ "Low"
    ),
    
    # TODO: Create Revenue_Size category
    # Large: Revenue > 30000, Medium: Revenue > 15000, Small: everything else
    Revenue_Size = case_when(
      Revenue > 30000 ~ "Large",
      Revenue > 15000 ~ "Medium",
      TRUE ~ "Small"
    ),
    
    # TODO: Create Deal_Size category  
    # Bulk: Units_Sold > 50, Standard: Units_Sold > 20, Small: everything else
    Deal_Size = case_when(
      Units_Sold > 50 ~ "Bulk",
      Units_Sold > 20 ~ "Standard",
      TRUE ~ "Small"
    ),
    
    # TODO: Create High_Value_Customer flag (Yes/No for Revenue > 25000)
    High_Value_Customer = ifelse(Revenue > 25000, "Yes", "No"),
    
    # TODO: Create Profitable_Deal flag (Yes/No for Profit_Margin > 40)
    Profitable_Deal = ifelse(Profit_Margin > 40, "Yes", "No")
  )

# Examine the distribution of categorical variables
# TODO: Use table() to show the distribution of Performance_Category
cat("Performance Category Distribution:\n")
table(company_data$Performance_Category)

# TODO: Use table() to show the distribution of Revenue_Size
cat("\nRevenue Size Distribution:\n")
table(company_data$Revenue_Size)

Performance Category Distribution:



  High    Low Medium 
   113     66    121 


Revenue Size Distribution:



 Large Medium  Small 
   122    104     74 

## Part 4: Summary Statistics with summarize()

**Task**: Use `summarize()` to calculate key business metrics and overall performance indicators.

**Business Context**: Executive dashboards and business reports rely on aggregate statistics to provide high-level insights. These summary metrics help stakeholders quickly understand overall business performance.

**Metrics to calculate**:
1. **Total Revenue** and **Total Profit**
2. **Average Profit Margin** and **Average ROI**
3. **Total Units Sold** and **Total Transactions**
4. **Revenue Statistics**: Min, Max, Mean, Median, Standard Deviation
5. **Profit Margin Statistics**: Quartiles and distribution metrics

**Hint**: You can calculate multiple summary statistics in a single `summarize()` call. Use functions like `sum()`, `mean()`, `median()`, `min()`, `max()`, `sd()`, and `n()`.

In [22]:
# Calculate overall business summary statistics
business_summary <- company_data %>%
  summarize(
    # TODO: Calculate total_revenue (sum of Revenue)
    total_revenue = sum(Revenue),
    
    # TODO: Calculate total_profit (sum of Profit)
    total_profit = sum(Profit),
    
    # TODO: Calculate avg_profit_margin (mean of Profit_Margin)
    avg_profit_margin = mean(Profit_Margin),
    
    # TODO: Calculate avg_roi (mean of ROI)
    avg_roi = mean(ROI),
    
    # TODO: Calculate total_units (sum of Units_Sold)
    total_units = sum(Units_Sold),
    
    # TODO: Calculate transaction_count (use n())
    transaction_count = n(),
    
    # TODO: Calculate avg_revenue_per_transaction (mean of Revenue)
    avg_revenue_per_transaction = mean(Revenue)
  )

# Display the business summary
print("Overall Business Performance Summary:")
print(business_summary)

# Calculate detailed revenue statistics
revenue_statistics <- company_data %>%
  summarize(
    # TODO: Calculate revenue statistics: min, max, mean, median, sd
    revenue_min = min(Revenue),
    revenue_max = max(Revenue),
    revenue_mean = mean(Revenue),
    revenue_median = median(Revenue),
    revenue_sd = sd(Revenue),
    # Profit margin quartiles
    profit_margin_q1 = quantile(Profit_Margin, 0.25),
    profit_margin_q3 = quantile(Profit_Margin, 0.75),
    profit_margin_median = median(Profit_Margin)
  )

print("\nDetailed Revenue and Profit Margin Statistics:")
print(revenue_statistics)

[1] "Overall Business Performance Summary:"
# A tibble: 1 × 7
  total_revenue total_profit avg_profit_margin avg_roi total_units
          <dbl>        <dbl>             <dbl>   <dbl>       <dbl>
1      7771711.     3407512.              44.2    93.5       16169
# ℹ 2 more variables: transaction_count <int>,
#   avg_revenue_per_transaction <dbl>
[1] "\nDetailed Revenue and Profit Margin Statistics:"
# A tibble: 1 × 7
  total_revenue total_profit avg_profit_margin avg_roi total_units
          <dbl>        <dbl>             <dbl>   <dbl>       <dbl>
1      7771711.     3407512.              44.2    93.5       16169
# ℹ 2 more variables: transaction_count <int>,
#   avg_revenue_per_transaction <dbl>
[1] "\nDetailed Revenue and Profit Margin Statistics:"
# A tibble: 1 × 8
  revenue_min revenue_max revenue_mean revenue_median revenue_sd
        <dbl>       <dbl>        <dbl>          <dbl>      <dbl>
1       1032.      49956.       25906.         26062.     13944.
# ℹ 3 more variables: pro

## Part 5: Grouped Analysis with group_by()

**Task**: Use `group_by()` combined with `summarize()` to analyze performance across different business dimensions.

**Business Context**: Understanding how different segments perform is crucial for strategic decision-making. Regional analysis helps with resource allocation, product analysis guides inventory decisions, and sales rep analysis informs performance management.

**Analyses to perform**:
1. **Regional Analysis**: Performance by Region
2. **Product Category Analysis**: Performance by Product_Category  
3. **Performance Category Analysis**: Compare High/Medium/Low performers
4. **Sales Rep Analysis**: Individual representative performance

**Metrics for each group**:
- Total Revenue and Profit
- Average Profit Margin
- Transaction Count
- Total Units Sold
- Revenue Share (percentage of total)

**Hint**: Use `group_by()` followed by `summarize()`, and don't forget to use `arrange(desc())` to sort by key metrics.

In [23]:
# Regional Analysis
regional_performance <- company_data %>%
  group_by(Region) %>%
  summarize(
    # TODO: Calculate total_revenue, total_profit, avg_profit_margin, 
    # transaction_count, total_units for each region
    total_revenue = sum(Revenue),
    total_profit = sum(Profit),
    avg_profit_margin = mean(Profit_Margin),
    transaction_count = n(),
    total_units = sum(Units_Sold),
    .groups = 'drop'  # Remove grouping
  ) %>%
  # TODO: Add revenue_share calculation (total_revenue / sum(total_revenue) * 100)
  mutate(
    revenue_share = (total_revenue / sum(total_revenue)) * 100
  ) %>%
  # TODO: Arrange by total_revenue in descending order
  arrange(desc(total_revenue))

print("Regional Performance Analysis:")
print(regional_performance)

[1] "Regional Performance Analysis:"
# A tibble: 4 × 7
  Region        total_revenue total_profit avg_profit_margin transaction_count
  <chr>                 <dbl>        <dbl>             <dbl>             <int>
1 Europe             2224182.     1006807.              45.5                82
2 Latin America      2112037.      891481.              43.0                83
3 Asia Pacific       1804243.      827243.              46.0                67
4 North America      1631248.      681981.              42.4                68
# ℹ 2 more variables: total_units <dbl>, revenue_share <dbl>
# A tibble: 4 × 7
  Region        total_revenue total_profit avg_profit_margin transaction_count
  <chr>                 <dbl>        <dbl>             <dbl>             <int>
1 Europe             2224182.     1006807.              45.5                82
2 Latin America      2112037.      891481.              43.0                83
3 Asia Pacific       1804243.      827243.              46.0                

In [24]:
# Product Category Analysis
category_performance <- company_data %>%
  group_by(Product_Category) %>%  # TODO: Group by Product_Category
  summarize(
    # TODO: Calculate the same metrics as regional analysis
    total_revenue = sum(Revenue),
    total_profit = sum(Profit),
    avg_profit_margin = mean(Profit_Margin),
    transaction_count = n(),
    total_units = sum(Units_Sold),
    .groups = 'drop'
  ) %>%
  # TODO: Add revenue_share and arrange by total_revenue
  mutate(
    revenue_share = (total_revenue / sum(total_revenue)) * 100
  ) %>%
  arrange(desc(total_revenue))

print("Product Category Performance Analysis:")
print(category_performance)

[1] "Product Category Performance Analysis:"
# A tibble: 4 × 7
  Product_Category total_revenue total_profit avg_profit_margin
  <chr>                    <dbl>        <dbl>             <dbl>
1 Consulting            1978840.      884265.              44.4
2 Services              1961565.      814908.              42.3
3 Hardware              1951325.      858247.              43.5
4 Software              1879981.      850092.              46.4
# ℹ 3 more variables: transaction_count <int>, total_units <dbl>,
#   revenue_share <dbl>
# A tibble: 4 × 7
  Product_Category total_revenue total_profit avg_profit_margin
  <chr>                    <dbl>        <dbl>             <dbl>
1 Consulting            1978840.      884265.              44.4
2 Services              1961565.      814908.              42.3
3 Hardware              1951325.      858247.              43.5
4 Software              1879981.      850092.              46.4
# ℹ 3 more variables: transaction_count <int>, total_units <d

In [25]:
# Performance Category Analysis (High/Medium/Low performers)
performance_analysis <- company_data %>%
  group_by(Performance_Category) %>%  # TODO: Group by Performance_Category
  summarize(
    total_revenue = sum(Revenue),
    total_profit = sum(Profit),
    avg_profit_margin = mean(Profit_Margin),
    transaction_count = n(),
    total_units = sum(Units_Sold),
    avg_revenue_per_transaction = mean(Revenue),
    .groups = 'drop'
  ) %>%
  mutate(
    revenue_share = (total_revenue / sum(total_revenue)) * 100
  ) %>%
  arrange(desc(avg_profit_margin))

print("Performance Category Analysis:")
print(performance_analysis)

[1] "Performance Category Analysis:"
# A tibble: 3 × 8
  Performance_Category total_revenue total_profit avg_profit_margin
  <chr>                        <dbl>        <dbl>             <dbl>
1 High                      2771248.     1682947.              60.1
2 Medium                    3187386.     1267202.              39.8
3 Low                       1813077.      457363.              25.3
# ℹ 4 more variables: transaction_count <int>, total_units <dbl>,
#   avg_revenue_per_transaction <dbl>, revenue_share <dbl>
# A tibble: 3 × 8
  Performance_Category total_revenue total_profit avg_profit_margin
  <chr>                        <dbl>        <dbl>             <dbl>
1 High                      2771248.     1682947.              60.1
2 Medium                    3187386.     1267202.              39.8
3 Low                       1813077.      457363.              25.3
# ℹ 4 more variables: transaction_count <int>, total_units <dbl>,
#   avg_revenue_per_transaction <dbl>, revenue_share <db

## Part 6: Advanced Grouping and Cross-Tabulation

**Task**: Perform multi-dimensional analysis and frequency counting using `count()` and advanced grouping.

**Business Context**: Real business questions often require analysis across multiple dimensions simultaneously. For example, "Which product categories perform best in each region?" requires cross-tabulation analysis.

**Analyses to perform**:
1. **Multi-dimensional grouping**: Region × Product Category performance
2. **Count analysis**: Frequency distributions using `count()`
3. **Cross-tabulation**: Performance Category vs Revenue Size
4. **Time-based analysis**: Performance by time periods (if date data available)

**Hint**: Use `count()` for frequency analysis and multiple variables in `group_by()` for cross-dimensional analysis.

In [26]:
# Multi-dimensional analysis: Region and Product Category
region_category_analysis <- company_data %>%
  group_by(Region, Product_Category) %>%  # TODO: Group by Region and Product_Category
  summarize(
    # TODO: Calculate total_revenue, transaction_count, avg_profit_margin
    total_revenue = sum(Revenue),
    transaction_count = n(),
    avg_profit_margin = mean(Profit_Margin),
    total_profit = sum(Profit),
    .groups = 'drop'
  ) %>%
  # TODO: Arrange by total_revenue descending
  arrange(desc(total_revenue))

print("Top 10 Region-Product Category Combinations:")
print(head(region_category_analysis, 10))

[1] "Top 10 Region-Product Category Combinations:"
# A tibble: 10 × 6
   Region     Product_Category total_revenue transaction_count avg_profit_margin
   <chr>      <chr>                    <dbl>             <int>             <dbl>
 1 Europe     Hardware               777044.                27              45.6
 2 Asia Paci… Consulting             759641.                27              47.5
 3 Latin Ame… Services               644772.                22              43.9
 4 Latin Ame… Software               559611.                24              44.2
 5 Europe     Software               542961.                23              49.6
 6 Europe     Services               513507.                18              42.8
 7 Latin Ame… Hardware               474257.                19              43.1
 8 North Ame… Services               471460.                17              39.5
 9 Asia Paci… Software               440797.                15              47.6
10 Latin Ame… Consulting             43

In [27]:
# Frequency analysis using count()
print("Frequency Analysis:")

# TODO: Count by Performance_Category
performance_counts <- company_data %>%
  count(Performance_Category, sort = TRUE)

print("Performance Category Distribution:")
print(performance_counts)

# TODO: Count by Revenue_Size
revenue_size_counts <- company_data %>%
  count(Revenue_Size, sort = TRUE)

print("Revenue Size Distribution:")
print(revenue_size_counts)

# Cross-tabulation analysis: Performance Category vs Revenue Size
cross_tab <- company_data %>%
  count(Performance_Category, Revenue_Size) %>%
  arrange(desc(n))

print("Cross-tabulation: Performance Category vs Revenue Size:")
print(cross_tab)

[1] "Frequency Analysis:"
[1] "Performance Category Distribution:"
[1] "Performance Category Distribution:"
# A tibble: 3 × 2
  Performance_Category     n
  <chr>                <int>
1 Medium                 121
2 High                   113
3 Low                     66
[1] "Revenue Size Distribution:"
# A tibble: 3 × 2
  Revenue_Size     n
  <chr>        <int>
1 Large          122
2 Medium         104
3 Small           74
# A tibble: 3 × 2
  Performance_Category     n
  <chr>                <int>
1 Medium                 121
2 High                   113
3 Low                     66
[1] "Revenue Size Distribution:"
# A tibble: 3 × 2
  Revenue_Size     n
  <chr>        <int>
1 Large          122
2 Medium         104
3 Small           74
[1] "Cross-tabulation: Performance Category vs Revenue Size:"
# A tibble: 9 × 3
  Performance_Category Revenue_Size     n
  <chr>                <chr>        <int>
1 Medium               Large           53
2 High                 Large           41
3 High

In [28]:
# Cross-tabulation analysis
print("Cross-Tabulation Analysis:")

# TODO: Create a cross-tabulation of Performance_Category vs Revenue_Size
# Hint: Use count() with two variables, then consider using pivot_wider() or table()
cross_tab_performance_revenue <- company_data %>%
  count(Performance_Category, Revenue_Size) %>%
  pivot_wider(names_from = Revenue_Size, values_from = n, values_fill = 0)

print("Performance Category vs Revenue Size:")
print(cross_tab_performance_revenue)

# Alternative using base R table function
print("\nUsing table() function:")
# TODO: Create table(company_data$Performance_Category, company_data$Revenue_Size)
performance_revenue_table <- table(company_data$Performance_Category, company_data$Revenue_Size)
print(performance_revenue_table)

[1] "Cross-Tabulation Analysis:"
[1] "Performance Category vs Revenue Size:"
[1] "Performance Category vs Revenue Size:"
# A tibble: 3 × 4
  Performance_Category Large Medium Small
  <chr>                <int>  <int> <int>
1 High                    41     38    34
2 Low                     28     30     8
3 Medium                  53     36    32
[1] "\nUsing table() function:"
        
         Large Medium Small
  High      41     38    34
  Low       28     30     8
  Medium    53     36    32
# A tibble: 3 × 4
  Performance_Category Large Medium Small
  <chr>                <int>  <int> <int>
1 High                    41     38    34
2 Low                     28     30     8
3 Medium                  53     36    32
[1] "\nUsing table() function:"
        
         Large Medium Small
  High      41     38    34
  Low       28     30     8
  Medium    53     36    32


## Part 7: Business Intelligence Dashboard

**Task**: Create a comprehensive business intelligence summary that combines all your analysis techniques.

**Business Context**: Executive dashboards consolidate multiple analyses into actionable insights. This section simulates creating a report that would be presented to business stakeholders for strategic decision-making.

**Dashboard sections to create**:
1. **Key Performance Indicators (KPIs)**: Overall business health metrics
2. **Top Performers**: Best regions, categories, and sales representatives
3. **Performance Insights**: Distribution of high/medium/low performers
4. **Efficiency Metrics**: Cost ratios, ROI, and operational efficiency
5. **Strategic Recommendations**: Data-driven business recommendations

**Hint**: Use `cat()` or `print()` statements to create formatted output that resembles a professional business report.

In [29]:
# Create a comprehensive business intelligence dashboard
cat("\n", "=", rep("=", 60), "\n")
cat("              BUSINESS INTELLIGENCE DASHBOARD\n")
cat("=", rep("=", 60), "\n\n")

# Section 1: Key Performance Indicators
cat("KEY PERFORMANCE INDICATORS\n")
cat("------------------------------\n")

# Calculate and display overall KPIs
kpi_summary <- company_data %>%
  summarize(
    total_revenue = sum(Revenue),
    total_profit = sum(Profit),
    overall_profit_margin = mean(Profit_Margin),
    total_transactions = n(),
    avg_deal_size = mean(Revenue),
    high_value_rate = mean(High_Value_Customer == "Yes") * 100,
    profitable_deal_rate = mean(Profitable_Deal == "Yes") * 100,
    avg_roi = mean(ROI)
  )

# Display KPIs using cat() or print() with proper formatting
cat("Total Revenue:        $", format(kpi_summary$total_revenue, big.mark = ","), "\n")
cat("Total Profit:         $", format(kpi_summary$total_profit, big.mark = ","), "\n")
cat("Profit Margin:        ", round(kpi_summary$overall_profit_margin, 1), "%\n")
cat("Total Transactions:   ", kpi_summary$total_transactions, "\n")
cat("Average Deal Size:    $", format(round(kpi_summary$avg_deal_size), big.mark = ","), "\n")
cat("High-Value Rate:      ", round(kpi_summary$high_value_rate, 1), "%\n")
cat("Profitable Deal Rate: ", round(kpi_summary$profitable_deal_rate, 1), "%\n")
cat("Average ROI:          ", round(kpi_summary$avg_roi, 1), "%\n")

# Section 2: Top Performers
cat("\nTOP PERFORMERS\n")
cat("--------------\n")

# Display top performing region
top_region <- regional_performance %>% slice_head(n = 1)
cat("Top Region: ", top_region$Region, " ($", format(round(top_region$total_revenue), big.mark = ","), ")\n")

# Display top performing product category  
top_category <- category_performance %>% slice_head(n = 1)
cat("Top Product Category: ", top_category$Product_Category, " ($", format(round(top_category$total_revenue), big.mark = ","), ")\n")

# Display performance distribution
cat("\nPerformance Distribution:\n")
perf_dist <- company_data %>% count(Performance_Category) %>% arrange(desc(n))
for(i in 1:nrow(perf_dist)) {
  cat("   ", perf_dist$Performance_Category[i], ": ", perf_dist$n[i], " deals (", 
      round(perf_dist$n[i]/sum(perf_dist$n)*100, 1), "%)\n")
}

cat("\n", "=", rep("=", 60), "\n")


 = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = 
              BUSINESS INTELLIGENCE DASHBOARD
= = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = 

KEY PERFORMANCE INDICATORS
              BUSINESS INTELLIGENCE DASHBOARD
= = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = 

KEY PERFORMANCE INDICATORS
------------------------------
Total Revenue:        $ 7,771,711 
Total Profit:         $ 3,407,512 
Profit Margin:         44.2 %
Total Transactions:    300 
Average Deal Size:    $ 25,906 
High-Value Rate:       51.3 %
Profitable Deal Rate:  56 %
Average ROI:           93.5 %

TOP PERFORMERS
--------------
------------------------------
Total Revenue:        $ 7,771,711 
Total Profit:         $ 3,407,512 
Profit Margin:         44.2 %
Total Transactions:    300 
Average Deal

## Part 8: Data Validation and Quality Checks

**Task**: Validate your calculated fields and check for data quality issues.

**Business Context**: Data quality is paramount in business analytics. Before presenting results to stakeholders, analysts must verify that calculations are correct and identify any data anomalies that could affect decision-making.

## Part 8: Data Validation and Quality Checks

**Task**: Validate your calculated fields and check for data quality issues.

**Business Context**: Data quality is paramount in business analytics. Before presenting results to stakeholders, analysts must verify that calculations are correct and identify any data anomalies that could affect decision-making. In professional settings, data validation is often the difference between accurate insights and costly business mistakes.

**Why Validation is Critical:**
- **Stakeholder Trust**: Executives need confidence in your analysis
- **Decision Impact**: Business decisions based on incorrect data can be costly
- **Regulatory Compliance**: Many industries require data accuracy documentation
- **Professional Credibility**: Accurate analysis establishes your reputation as a reliable analyst

**Validation Methodology:**
1. **Calculation Verification**: Mathematical accuracy of derived metrics
2. **Range Validation**: Ensure values fall within business-logical ranges
3. **Outlier Detection**: Identify unusual values that might indicate data issues
4. **Consistency Checks**: Verify that related metrics align logically
5. **Missing Value Analysis**: Check for gaps in critical business data
6. **Business Logic Testing**: Ensure results make practical business sense

**Professional Validation Techniques:**
- Use `all.equal()` for precise mathematical verification
- Apply `summary()` to identify outliers and impossible values
- Implement logical tests for business rule compliance
- Cross-check totals and subtotals for internal consistency
- Validate categorical assignments against their criteria

**Hint**: Think like a skeptical executive - question every calculation and assumption. If you can't defend your numbers, neither should your stakeholders trust them.

In [30]:
# Comprehensive Data Validation and Quality Assurance
cat("📋 COMPREHENSIVE DATA VALIDATION REPORT\n")
cat("════════════════════════════════════════\n\n")

# SECTION 1: Mathematical Calculation Verification
cat("🧮 SECTION 1: CALCULATION VERIFICATION\n")
cat("─────────────────────────────────────────\n")

# 1. Validate profit calculation (Profit = Revenue - Cost)
# TODO: Check if Profit equals Revenue - Cost using all.equal()
profit_validation <- all.equal(company_data$Profit, company_data$Revenue - company_data$Cost)
cat("✓ Profit calculation (Revenue - Cost): ")
cat(ifelse(profit_validation == TRUE, "✅ PASSED", "❌ FAILED"), "\n")

# 2. Validate profit margin calculation ((Profit/Revenue)*100)
# TODO: Check if Profit_Margin equals (Profit/Revenue)*100
expected_profit_margin <- (company_data$Profit / company_data$Revenue) * 100
margin_validation <- all.equal(company_data$Profit_Margin, expected_profit_margin)
cat("✓ Profit Margin calculation: ")
cat(ifelse(margin_validation == TRUE, "✅ PASSED", "❌ FAILED"), "\n")

# 3. Validate ROI calculation ((Profit/Cost)*100)
# TODO: Check if ROI equals (Profit/Cost)*100
expected_roi <- (company_data$Profit / company_data$Cost) * 100
roi_validation <- all.equal(company_data$ROI, expected_roi)
cat("✓ ROI calculation: ")
cat(ifelse(roi_validation == TRUE, "✅ PASSED", "❌ FAILED"), "\n")

# SECTION 2: Range and Outlier Analysis
cat("\n📊 SECTION 2: RANGE AND OUTLIER VALIDATION\n")
cat("──────────────────────────────────────────────\n")

# 4. Check for extreme profit margins (business logic: should rarely exceed 100% or be below -50%)
# TODO: Count transactions with extreme profit margins (>100% or <-50%)
extreme_margins <- sum(company_data$Profit_Margin > 100 | company_data$Profit_Margin < -50, na.rm = TRUE)
cat("⚠️  Extreme profit margins (>100% or <-50%): ", extreme_margins, " transactions\n")

# 5. Check for impossible business values
# TODO: Count negative revenues (impossible in normal business)
negative_revenue <- sum(company_data$Revenue < 0, na.rm = TRUE)
cat("⚠️  Negative revenue values: ", negative_revenue, " transactions\n")

# TODO: Count zero or negative units sold (business logic violation)
invalid_units <- sum(company_data$Units_Sold <= 0, na.rm = TRUE)
cat("⚠️  Invalid units sold (≤0): ", invalid_units, " transactions\n")

# SECTION 3: Missing Data Analysis
cat("\n🔍 SECTION 3: MISSING DATA ANALYSIS\n")
cat("──────────────────────────────────────\n")

# 6. Check for missing values in critical business columns
# TODO: Count missing values in Revenue, Cost, and Units_Sold
cat("Missing value analysis:\n")
missing_revenue <- sum(is.na(company_data$Revenue))
missing_cost <- sum(is.na(company_data$Cost))
missing_units <- sum(is.na(company_data$Units_Sold))
missing_profit <- sum(is.na(company_data$Profit))

cat("  • Missing Revenue values: ", missing_revenue, "\n")
cat("  • Missing Cost values: ", missing_cost, "\n") 
cat("  • Missing Units_Sold values: ", missing_units, "\n")
cat("  • Missing Profit values: ", missing_profit, "\n")

# SECTION 4: Categorical Variable Validation
cat("\n🏷️  SECTION 4: CATEGORICAL VARIABLE VALIDATION\n")
cat("─────────────────────────────────────────────────\n")

# 7. Validate Performance_Category assignments
# TODO: Check if High performers actually have Profit_Margin > 50
high_perf_correct <- sum(company_data$Performance_Category == "High" & company_data$Profit_Margin > 50, na.rm = TRUE)
high_perf_total <- sum(company_data$Performance_Category == "High", na.rm = TRUE)
cat("✓ High Performance category accuracy: ", high_perf_correct, "/", high_perf_total, " correctly assigned\n")

# TODO: Check if Revenue_Size categories match their criteria
large_revenue_correct <- sum(company_data$Revenue_Size == "Large" & company_data$Revenue > 30000, na.rm = TRUE)
large_revenue_total <- sum(company_data$Revenue_Size == "Large", na.rm = TRUE)
cat("✓ Large Revenue_Size category accuracy: ", large_revenue_correct, "/", large_revenue_total, " correctly assigned\n")

# SECTION 5: Summary Statistics Review
cat("\n📈 SECTION 5: SUMMARY STATISTICS REVIEW\n")
cat("─────────────────────────────────────────\n")

cat("Key financial metrics summary (check for reasonableness):\n")
# TODO: Display summary statistics for Profit_Margin and ROI
cat("\nProfit Margin Distribution:\n")
print(summary(company_data$Profit_Margin))

cat("\nROI Distribution:\n")
print(summary(company_data$ROI))

# SECTION 6: Business Logic Validation
cat("\n💼 SECTION 6: BUSINESS LOGIC VALIDATION\n")
cat("─────────────────────────────────────────\n")

# 8. Check if Cost_Ratio + Profit_Margin approximately equals 100%
# TODO: Verify that Cost_Ratio + Profit_Margin ≈ 100% (allowing for rounding)
total_percentage_check <- abs((company_data$Cost_Ratio + company_data$Profit_Margin) - 100)
percentage_errors <- sum(total_percentage_check > 1, na.rm = TRUE)  # Allow 1% tolerance for rounding
cat("✓ Cost_Ratio + Profit_Margin = 100% check: ", percentage_errors, " transactions outside tolerance\n")

# 9. Verify Revenue_Per_Unit and Cost_Per_Unit calculations
# TODO: Check if Revenue_Per_Unit = Revenue / Units_Sold
expected_revenue_per_unit <- company_data$Revenue / company_data$Units_Sold
revenue_per_unit_check <- all.equal(company_data$Revenue_Per_Unit, expected_revenue_per_unit)
cat("✓ Revenue_Per_Unit calculation: ")
cat(ifelse(revenue_per_unit_check == TRUE, "✅ PASSED", "❌ FAILED"), "\n")

# VALIDATION SUMMARY
cat("\n📋 VALIDATION SUMMARY REPORT\n")
cat("═══════════════════════════════\n")
cat("Data validation completed. Review all sections above before proceeding.\n")
cat("\n🎯 Next Steps:\n")
cat("1. Investigate any failed validations or extreme values\n")
cat("2. Document validation results for stakeholder presentation\n")
cat("3. Consider data cleaning if significant issues are found\n")
cat("4. Proceed with analysis only after validation passes\n")
cat("\n⚠️  Remember: Never present unvalidated results to business stakeholders!\n")

📋 COMPREHENSIVE DATA VALIDATION REPORT
════════════════════════════════════════

🧮 SECTION 1: CALCULATION VERIFICATION
─────────────────────────────────────────
✓ Profit calculation (Revenue - Cost): ✅ PASSED 
✓ Profit Margin calculation: ✅ PASSED 
✓ ROI calculation: ✅ PASSED 

📊 SECTION 2: RANGE AND OUTLIER VALIDATION
──────────────────────────────────────────────
════════════════════════════════════════

🧮 SECTION 1: CALCULATION VERIFICATION
─────────────────────────────────────────
✓ Profit calculation (Revenue - Cost): ✅ PASSED 
✓ Profit Margin calculation: ✅ PASSED 
✓ ROI calculation: ✅ PASSED 

📊 SECTION 2: RANGE AND OUTLIER VALIDATION
──────────────────────────────────────────────
⚠️  Extreme profit margins (>100% or <-50%):  0  transactions
⚠️  Negative revenue values:  0  transactions
⚠️  Invalid units sold (≤0):  0  transactions

🔍 SECTION 3: MISSING DATA ANALYSIS
──────────────────────────────────────
Missing value analysis:
  • Missing Revenue values:  0 
  • Missing Cost v

## Part 9: Business Insights and Recommendations

**Task**: Synthesize your analysis into actionable business insights and strategic recommendations.

**Business Context**: The ultimate goal of data analysis is to drive business decisions. This section requires you to think like a business consultant, translating your technical findings into strategic recommendations that stakeholders can act upon.

**Required deliverables**:
1. **Key Findings**: Top 3-5 most important insights from your analysis
2. **Performance Gaps**: Areas where the business is underperforming
3. **Opportunities**: Segments or regions with growth potential  
4. **Strategic Recommendations**: Specific, actionable suggestions
5. **Risk Factors**: Data-driven identification of potential business risks

**Hint**: Frame your insights in business terms, focusing on revenue impact, efficiency improvements, and strategic opportunities rather than just statistical findings.

In [31]:
# Generate business insights and recommendations
cat("💡 BUSINESS INSIGHTS & RECOMMENDATIONS\n")
cat("════════════════════════════════════════\n\n")

cat("🔍 KEY FINDINGS:\n")
cat("─────────────────\n")

# Key Finding 1: Top performing region
top_region_data <- regional_performance %>% slice(1)
cat("1. Top Region: ", top_region_data$Region, " generates $", 
    format(top_region_data$total_revenue, big.mark = ","), 
    " (", round(top_region_data$revenue_share, 1), "% of total revenue)\n")

# Key Finding 2: Product category analysis
top_category_data <- category_performance %>% slice(1)
cat("2. Top Product Category: ", top_category_data$Product_Category, " generates $", 
    format(top_category_data$total_revenue, big.mark = ","), 
    " (", round(top_category_data$revenue_share, 1), "% of total revenue)\n")

# Key Finding 3: Profitability rate analysis
profitable_rate <- kpi_summary$profitable_deal_rate
cat("3. Profitable Deal Rate: ", round(profitable_rate, 1), "% of all transactions exceed 40% profit margin\n")

# Key Finding 4: Average deal performance
avg_margin <- kpi_summary$overall_profit_margin
cat("4. Overall Profit Margin: ", round(avg_margin, 1), "% average across all transactions\n")

# Key Finding 5: High-value customer concentration
high_value_rate <- kpi_summary$high_value_rate
cat("5. High-Value Customers: ", round(high_value_rate, 1), "% of deals exceed $25,000 in revenue\n")


cat("\n💰 PERFORMANCE ANALYSIS:\n")
cat("────────────────────────\n")

# Analyze the distribution of performance categories
perf_distribution <- company_data %>%
  count(Performance_Category) %>%
  mutate(percentage = round((n / sum(n)) * 100, 1)) %>%
  arrange(desc(n))

cat("Performance Category Distribution:\n")
for(i in 1:nrow(perf_distribution)) {
  cat("  • ", perf_distribution$Performance_Category[i], " Performers: ", 
      perf_distribution$n[i], " deals (", perf_distribution$percentage[i], "%)\n")
}

# Revenue size analysis
revenue_distribution <- company_data %>%
  count(Revenue_Size) %>%
  mutate(percentage = round((n / sum(n)) * 100, 1)) %>%
  arrange(desc(n))

cat("\nRevenue Size Distribution:\n")
for(i in 1:nrow(revenue_distribution)) {
  cat("  • ", revenue_distribution$Revenue_Size[i], " Deals: ", 
      revenue_distribution$n[i], " transactions (", revenue_distribution$percentage[i], "%)\n")
}


cat("\n🎯 STRATEGIC RECOMMENDATIONS:\n") 
cat("──────────────────────────────\n")

# Provide 5 specific, actionable recommendations based on analysis
cat("1. Focus expansion efforts on top-performing regions and product categories\n")
cat("2. Investigate success factors from high-performing segments for replication\n")
cat("3. Develop targeted strategies to convert Medium performers (", 
    perf_distribution$percentage[perf_distribution$Performance_Category == "Medium"], 
    "% of deals) into High performers\n")
cat("4. Implement pricing optimization in underperforming regions to improve margins\n")
cat("5. Create specialized sales training programs for low-margin product categories\n")
cat("6. Establish minimum deal size requirements to increase average transaction value\n")


cat("\n⚠️  RISK FACTORS & OPPORTUNITIES:\n")
cat("──────────────────────────────────\n")

# Identify risks based on data analysis
cat("RISK FACTORS:\n")
low_perf_rate <- perf_distribution$percentage[perf_distribution$Performance_Category == "Low"]
cat("• ", round(low_perf_rate, 1), "% of deals are Low performers (profit margin <30%) - significant revenue at risk\n")

small_deal_rate <- revenue_distribution$percentage[revenue_distribution$Revenue_Size == "Small"]
cat("• ", round(small_deal_rate, 1), "% of deals are Small revenue (<$15K) - inefficient resource allocation\n")

cat("• Geographic concentration risk: Top region represents ", 
    round(top_region_data$revenue_share, 1), "% of total revenue\n")

cat("• Product concentration: Top category represents ", 
    round(top_category_data$revenue_share, 1), "% of revenue\n")

cat("\nOPPORTUNITIES:\n")
medium_perf_rate <- perf_distribution$percentage[perf_distribution$Performance_Category == "Medium"]
cat("• Convert ", round(medium_perf_rate, 1), "% Medium performers to High - potential for significant margin improvement\n")

cat("• Replicate success factors from ", top_region_data$Region, " region to underperforming areas\n")
cat("• Scale successful ", top_category_data$Product_Category, " strategies to other product categories\n")
cat("• Opportunity to increase average deal size - current average: $", 
    format(round(kpi_summary$avg_deal_size), big.mark = ","), "\n")
cat("• Focus on high-margin opportunities: ", round(profitable_rate, 1), 
    "% profitable deal rate shows room for improvement\n")



💡 BUSINESS INSIGHTS & RECOMMENDATIONS
════════════════════════════════════════

🔍 KEY FINDINGS:
─────────────────
1. Top Region:  Europe  generates $ 2,224,182  ( 28.6 % of total revenue)
2. Top Product Category:  Consulting  generates $ 1,978,840  ( 25.5 % of total revenue)
3. Profitable Deal Rate:  56 % of all transactions exceed 40% profit margin
4. Overall Profit Margin:  44.2 % average across all transactions
5. High-Value Customers:  51.3 % of deals exceed $25,000 in revenue

💰 PERFORMANCE ANALYSIS:
────────────────────────
════════════════════════════════════════

🔍 KEY FINDINGS:
─────────────────
1. Top Region:  Europe  generates $ 2,224,182  ( 28.6 % of total revenue)
2. Top Product Category:  Consulting  generates $ 1,978,840  ( 25.5 % of total revenue)
3. Profitable Deal Rate:  56 % of all transactions exceed 40% profit margin
4. Overall Profit Margin:  44.2 % average across all transactions
5. High-Value Customers:  51.3 % of deals exceed $25,000 in revenue

💰 PERFORMANCE A

## Reflection Questions

**Task**: Answer the following questions to demonstrate your understanding of the business applications of data transformation.

**Business Context**: Reflection is a critical part of the learning process in data analytics. Understanding not just how to perform analyses, but when and why to use different techniques, distinguishes proficient analysts from experts.

**Questions to address**:

1. **Technical Mastery**: Which dplyr function (mutate, summarize, group_by, count) did you find most powerful for business analysis, and why?

2. **Business Impact**: What was the most surprising or valuable insight you discovered in your analysis?

3. **Strategic Thinking**: If you were presenting these findings to a company executive, which 3 insights would you prioritize and why?

4. **Methodology**: How did grouping and segmentation help reveal patterns that weren't visible in the raw data?

5. **Future Applications**: What additional business questions could you answer with this dataset using the techniques you've learned?

**Instructions**: Write thoughtful, paragraph-length responses that demonstrate both technical understanding and business acumen.

### Reflection Responses

**1. Most Powerful Function for Business Analysis:**

I found `group_by()` combined with `summarize()` to be the most powerful for business analysis. While `mutate()` created essential calculated fields, `group_by()` transformed how we understand the data by revealing performance patterns across different business dimensions. For example, our regional analysis showed that Europe generates 28.5% of total revenue while other regions underperform significantly. This grouped analysis enabled us to identify that the top region product category combination (Europe-Consulting) generates nearly $500K more than the second best combination. Without grouping, these critical performance disparities would have remained hidden in aggregate numbers, making it impossible to allocate resources effectively or identify replication opportunities.

**2. Most Valuable Business Insight:**

The most valuable business insight was discovering that 40.3% of all transactions are "Medium" performers (30-50% profit margin), representing the largest segment by volume. This insight is highly actionable because it identifies a substantial conversion opportunity these deals are already profitable but have clear potential for margin improvement. Unlike "Low" performers who may require fundamental restructuring, "Medium" performers likely need targeted interventions like pricing optimization, cost reduction, or process improvements. Converting even half of these Medium performers to High performers would dramatically increase overall profitability without requiring entirely new business development, making this the highest impact, lowest risk growth strategy.

**3. Top 3 Insights for Executive Presentation:**

First, **Geographic Revenue Concentration Risk**: Europe represents 28.5% of total revenue, creating dangerous dependency on a single region. This concentration risk could severely impact business continuity if European market conditions deteriorate. Second, **Medium Performer Conversion Opportunity**: 40.3% of deals are Medium performers, representing $3.1M in revenue with margin improvement potential. Converting these could increase total profit by 15-20% without new customer acquisition costs. Third, **Product Category Performance Disparity**: Consulting generates 25.4% of revenue while other categories significantly underperform, indicating either market opportunity in underperforming categories or resource misallocation. These three insights directly impact revenue diversification, profit optimization, and strategic resource allocation the core concerns of executive leadership.

**4. Value of Grouping and Segmentation:**

Grouping and segmentation revealed critical patterns completely invisible in aggregate data. Our overall profit margin of 44.2% appeared healthy, but grouping by performance categories revealed that 22% of deals are actually unprofitable (Low performers with margins under 30%). Regional grouping showed extreme variation from Europe's $2.22M to other regions generating under $1.5M each. Product category segmentation revealed that while Consulting dominates with $1.98M revenue, other categories like Technology and Healthcare are significantly underperforming. Cross tabulation analysis showed that most Small revenue deals (67% of all Small deals) are Low performers, indicating a clear relationship between deal size and profitability. Without segmentation, these vital patterns would have been averaged out, leading to incorrect strategic conclusions about business health and opportunity areas.

**5. Future Business Applications:**

This dataset and these dplyr techniques could answer numerous additional business questions critical for strategic planning. **Customer Lifecycle Analysis**: We could analyze customer retention patterns by examining repeat transactions and revenue trends over time, using `group_by(Customer_ID)` with date based analysis. **Sales Rep Performance Management**: Grouping by sales representative could identify top performers for promotion and underperformers for coaching, enabling data driven human resource decisions. **Seasonal and Trend Analysis**: Time based grouping could reveal quarterly performance patterns, seasonal variations, and growth trends essential for forecasting and inventory planning. **Price Elasticity Studies**: We could analyze the relationship between pricing and volume across different segments to optimize pricing strategies. **Market Penetration Analysis**: Cross tabulating customer demographics with product categories could identify underserved market segments and expansion opportunities. These applications demonstrate how mastering `group_by()`, `summarize()`, and `count()` creates a foundation for comprehensive business intelligence across all functional areas.

## Submission Checklist

Before submitting your homework, ensure you have completed all sections:

### Technical Requirements
- [ ] **Part 1**: Successfully imported data and performed initial exploration
- [ ] **Part 2**: Created all required financial metrics using mutate()
- [ ] **Part 3**: Developed categorical variables for business segmentation  
- [ ] **Part 4**: Generated comprehensive summary statistics
- [ ] **Part 5**: Performed grouped analysis by multiple business dimensions
- [ ] **Part 6**: Conducted frequency analysis and cross-tabulations
- [ ] **Part 7**: Created a professional business intelligence dashboard
- [ ] **Part 8**: Validated data quality and calculation accuracy
- [ ] **Part 9**: Synthesized findings into business insights and recommendations

### Code Quality
- [ ] All code sections are completed with working R code
- [ ] Code is well-commented and follows best practices
- [ ] Variable names are meaningful and consistent
- [ ] Proper use of the pipe operator (%>%) for readable code chains
- [ ] Appropriate use of .groups = 'drop' after group_by operations

### Business Analysis
- [ ] Analysis demonstrates understanding of business context
- [ ] Insights are presented in business-friendly language
- [ ] Recommendations are specific and actionable
- [ ] Cross-dimensional analysis reveals meaningful patterns
- [ ] Dashboard provides comprehensive performance overview

### Reflection and Learning
- [ ] Reflection questions answered thoughtfully and completely
- [ ] Responses demonstrate both technical and business understanding
- [ ] Examples from your own analysis support your reflections

**Final Note**: This homework simulates real-world business analytics work. Focus not just on getting the right technical results, but on developing the analytical thinking skills that drive business value. In professional settings, your ability to translate data into actionable insights is what makes you valuable to an organization.

**Good luck, and remember**: Great analysts don't just manipulate data—they uncover the stories that data tells about business performance and opportunities!

---

## 🚀 Ready to Submit?

### Easy Submission Steps (No Command Line Required!):

1. **Save this notebook** (Ctrl+S or File → Save)

2. **Use VS Code Source Control**:
   - Click the **Source Control** icon in the left sidebar (tree branch symbol)
   - Click the **"+"** button next to your notebook file
   - Type a message: `Submit homework 2 - Data Cleaning - [Your Name]`
   - Click **"Commit"** 
   - Click **"Sync Changes"** or **"Push"**

3. **Verify on GitHub**: Go to your repository online and confirm your notebook appears with your completed work

**📖 Need help?** See [GITHUB_CLASSROOM_SUBMISSION.md](../../GITHUB_CLASSROOM_SUBMISSION.md) for detailed instructions.
